File: graphs/bfs_dfs.py
# simple bfs and dfs

def dfs_iterative(graph, start):
    # stack
    stack = [ start ]
    while (len(stack)):
        curr_node = stack.pop()
        for neighbour in graph[curr_node]:
            stack.append(neighbour)
        print(f"curr_node: {curr_node}")
    
def dfs_recursive(graph, start):
    # call stack
    print(f"curr_node: {start}")
    if len(graph[start]) == 0: return
    for neighbour in graph[start]:
        dfs_recursive(graph, neighbour)



def bfs(graph, start):
    # queue
    queue = [ start ]
    while (len(queue)):
        curr_node = queue.pop()
        for neighbour in graph[curr_node]:
            queue.insert(0, neighbour)
        print(f"curr_node: {curr_node}")
      
def main():
    # graph
    graph = {
             "a" : ["b", "c"],
             "b" : ["d"],
             "c" : ["e"],
             "d" : ["f"],
             "e" : [],
             "f" : []
            }
    
    print("DFS Iterative")
    dfs_iterative(graph, "a")
    
    print("DFS Recursive")
    dfs_recursive(graph, "a")

    print("BFS")
    bfs(graph, "a")
main()

File: graphs/bfs_dfs.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>

using namespace std;

void bfs(unordered_map< string, vector< string > >& graph,
             string src)
{
    vector< string > queue{ src };

    while (!queue.empty())
    {
        string currNode = queue.back();
        queue.pop_back();

        for (auto& neighbour : graph.at(currNode))
            queue.insert(queue.begin(), neighbour);
    
        cout << "currNode : " << currNode << "\n";
    }
}

void dfsRec(unordered_map< string, vector< string > >& graph,
             string src)
{
    
    cout << "currNode: " << src << "\n";

    for (auto& neighbour : graph.at(src))
        dfsRec(graph, neighbour);
    
}

void dfsIter(unordered_map< string, vector< string > >& graph,
             string src)
{
    vector < string > stack { src };

    while (!stack.empty())
    {
        string currNode = stack.back();
        stack.pop_back();
        for (auto& neighbour : graph.at(currNode))
            stack.push_back(neighbour);

        cout << "currNode: " << currNode << "\n";
    }
}

int main()
{
    unordered_map< string, vector< string > > graph = 
    {
        { "a" , vector< string >{ "b", "c" } },
        { "b" , vector< string >{ "d" } },
        { "c" , vector< string >{ "e" } },
        { "d" , vector< string >{ "f" } },
        { "e" , vector< string >{ } },
        { "f" , vector< string >{ } }
    };

    dfsIter(graph, "a");
    cout << "\n";
    dfsRec(graph, "a");
    cout << "\n";
    bfs(graph, "a");
    return 0;
}